ethcore/private-tx/src/log.rs

```diff
@@ -0,0 +1,409 @@
+	/// Validation timestamp, None if the transaction is not validated
+	pub validation_timestamp: Option<SystemTime>,
+}
+
+#[cfg(test)]
+impl PartialEq for ValidatorLog {
+	fn eq(&self, other: &Self) -> bool {
+		self.account == other.account
+	}
+}
+
+/// Information about the private transaction
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct TransactionLog {
+	/// Original signed transaction hash (used as a source for private tx)
+	pub tx_hash: H256,
+	/// Current status of the private transaction
+	pub status: PrivateTxStatus,
+	/// Creation timestamp
+	pub creation_timestamp: SystemTime,
+	/// List of validations
+	pub validators: Vec<ValidatorLog>,
+	/// Timestamp of the resulting public tx deployment
+	pub deployment_timestamp: Option<SystemTime>,
+	/// Hash of the resulting public tx
+	pub public_tx_hash: Option<H256>,
+}
+
+#[cfg(test)]
+impl PartialEq for TransactionLog {
+	fn eq(&self, other: &Self) -> bool {
+		self.tx_hash == other.tx_hash &&
+		self.status == other.status &&
+		self.validators == other.validators &&
+		self.public_tx_hash == other.public_tx_hash
+	}
+}
+
+/// Wrapper other JSON serializer
+pub trait LogsSerializer: Send + Sync + 'static {
+	/// Read logs from the source
+	fn read_logs(&self) -> Result<Vec<TransactionLog>, Error>;
+
+	/// Write all logs to the source
+	fn flush_logs(&self, logs: &HashMap<H256, TransactionLog>) -> Result<(), Error>;
+}
+
+pub struct FileLogsSerializer {
+	logs_dir: PathBuf,
+}
+
+impl FileLogsSerializer {
+	pub fn with_path<P: Into<PathBuf>>(logs_dir: P) -> Self {
+		FileLogsSerializer {
+			logs_dir: logs_dir.into(),
+		}
+	}
+
+	fn open_file(&self, to_create: bool) -> Result<File, Error> {
+		let file_path = self.logs_dir.with_file_name("private_tx.log");
+		if to_create {
+			File::create(&file_path).map_err(From::from)
+		} else {
+			File::open(&file_path).map_err(From::from)
+		}
+	}
+}
+
+impl LogsSerializer for FileLogsSerializer {
+	fn read_logs(&self) -> Result<Vec<TransactionLog>, Error> {
+		let log_file = self.open_file(false)?;
+		let transaction_logs: Vec<TransactionLog> = match serde_json::from_reader(log_file) {
+			Ok(logs) => logs,
+			Err(err) => {
+				error!(target: "privatetx", "Cannot deserialize logs from file: {}", err);
+				return Err(format!("Cannot deserialize logs from file: {:?}", err).into());
+			}
+		};
+		Ok(transaction_logs)
+	}
+
+	fn flush_logs(&self, logs: &HashMap<H256, TransactionLog>) -> Result<(), Error> {
+		if logs.is_empty() {
+			// Do not create empty file
+			return Ok(());
+		}
+		let log_file = self.open_file(true)?;
+		let mut json = serde_json::Serializer::new(log_file);
+		let mut json_array = json.serialize_seq(Some(logs.len()))?;
+		for v in logs.values() {
+			json_array.serialize_element(v)?;
+		}
+		json_array.end()?;
+		Ok(())
+	}
+}
+
+/// Private transactions logging
+pub struct Logging {
+	logs: RwLock<HashMap<H256, TransactionLog>>,
+	logs_serializer: Arc<LogsSerializer>,
+	mono_time: Mutex<MonoTime>,
```

Comment:
Why is this wrapped in a mutex? 

I mean it is never mutated AFAIU and just reads `Instant::elapsed()` or `Instand::elapsed() + start_time`?

Reply:
I think the Mutex is not needed. The existing`RwLock` implies sufficient synchronisation, i.e. all calls to `MonoTime` are guarded by a `logs.write()`.